hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/TestAllocateLatenciesMetrics.java

```diff
@@ -0,0 +1,260 @@
+  public void createAndStartRM() {
+    CapacitySchedulerConfiguration csConf =
+        new CapacitySchedulerConfiguration();
+    YarnConfiguration conf = new YarnConfiguration(csConf);
+    conf.setClass(YarnConfiguration.RM_SCHEDULER, CapacityScheduler.class,
+        ResourceScheduler.class);
+    conf.setBoolean(
+        YarnConfiguration.OPPORTUNISTIC_CONTAINER_ALLOCATION_ENABLED, true);
+    startRM(conf);
+  }
+
+  private void startRM(final YarnConfiguration conf) {
+    dispatcher = new DrainDispatcher();
+    rm = new MockRM(conf) {
+      @Override protected Dispatcher createDispatcher() {
+        return dispatcher;
+      }
+    };
+    rm.start();
+  }
+
+  @After
+  public void stopRM() {
+    if (rm != null) {
+      rm.stop();
+    }
+  }
+
+  @Test
+  public void testGuaranteedLatenciesMetrics() throws Exception {
+    HashMap<NodeId, MockNM> nodes = new HashMap<>();
+    MockNM nm1 = new MockNM("h1:1234", 4096, rm.getResourceTrackerService());
+    nodes.put(nm1.getNodeId(), nm1);
+    MockNM nm2 = new MockNM("h2:1234", 4096, rm.getResourceTrackerService());
+    nodes.put(nm2.getNodeId(), nm2);
+    nm1.registerNode();
+    nm2.registerNode();
+
+    nm1.nodeHeartbeat(oppContainersStatus, true);
+    nm2.nodeHeartbeat(oppContainersStatus, true);
+
+    MockRMAppSubmissionData data =
+        MockRMAppSubmissionData.Builder.createWithMemory(1 * GB, rm)
+            .withAppName("app").withUser("user").withAcls(null)
+            .withQueue("default").build();
+
+    RMApp app1 = MockRMAppSubmitter.submit(rm, data);
+    MockAM am1 = MockRM.launchAndRegisterAM(app1, rm, nm2);
+    ResourceScheduler scheduler = rm.getResourceScheduler();
+
+    // All nodes 1 to 2 will be applicable for scheduling.
+    nm1.nodeHeartbeat(oppContainersStatus, true);
+    nm2.nodeHeartbeat(oppContainersStatus, true);
+
+    ResourceRequest rr1 = ResourceRequest
+        .newInstance(Priority.newInstance(1), "*",
+            Resources.createResource(1 * GB), 1, true, null,
+            ExecutionTypeRequest.newInstance(ExecutionType.GUARANTEED, true));
+    ResourceRequest rr2 = ResourceRequest
+        .newInstance(Priority.newInstance(1), "*",
+            Resources.createResource(1 * GB), 1, true, null,
+            ExecutionTypeRequest.newInstance(ExecutionType.GUARANTEED, true));
+    rr1.setAllocationRequestId(1);
+    rr2.setAllocationRequestId(2);
+    AllocateResponse allocateResponse =
+        am1.allocate(Arrays.asList(rr1, rr2), null);
+
+    // make sure the containers get allocated
+    nm1.nodeHeartbeat(true);
+    List<Container> allocatedContainers =
+        am1.allocate(new ArrayList<ResourceRequest>(),
+            new ArrayList<ContainerId>()).getAllocatedContainers();
+    while (allocatedContainers.size() != 2) {
+      nm1.nodeHeartbeat(true);
+      allocatedContainers.addAll(am1.allocate(new ArrayList<ResourceRequest>(),
+          new ArrayList<ContainerId>()).getAllocatedContainers());
+      Thread.sleep(100);
```

Comment:
You could replace this with a waitFor so we bound the time instead of getting to a potential infinite loop.

Reply:
I've added maxRetries variable to avoid potential infinite loop.